Improve CLI robustness, safety, and readability

KushalMeghani1644 opened the pull request:
This PR makes several critical improvements to the Tauri CLI to enhance stability, performance, and maintainability:

1. **Thread safety and resilience:** The stdout/stderr reader threads now lock the mutex only per iteration, and poisoned locks are handled gracefully. This prevents deadlocks and panics when running commands concurrently, especially on long-running build processes.

2. **Performance and clarity:** Command argument formatting was refactored from fold to join, reducing memory allocations and improving readability.

3. **Cross-platform safety:** File extension parsing now correctly handles `OsStr` conversions, preventing potential panics or incorrect parsing on non-UTF-8 paths.

4. **Verbosity consistency:** Introduced a helper function to centralize verbosity parsing from environment variables and CLI flags, ensuring predictable logging levels across all subcommands.

5. **Default values:** Added `Default` derivations to config-related structs to make deserialization more robust and prevent unnecessary unwraps.

6. **Maintainability:** Standardized error handling and logging formatting for clearer messages and easier future modifications.

7. **Regression safety:** Added a minimal CLI help test to catch accidental changes to subcommand names or descriptions.

These changes are necessary to ensure the CLI behaves reliably under all conditions, handles errors gracefully, and remains maintainable as Tauri continues to grow.

KushalMeghani1644 committed:
Improve CLI robustness, safety, and readability

- Ensure granular mutex locking in stdout/stderr threads to prevent deadlocks and panics.
- Handle poisoned locks gracefully to improve CLI resilience.
- Replace inefficient fold-based command argument formatting with join for better performance.
- Simplify file extension handling using and_then(|ext| ext.to_str()) for clarity and cross-platform safety.
- Introduce get_verbosity helper for consistent and maintainable verbosity handling.
- Add Default derives to configuration structs for safer deserialization.
- Add minimal CLI smoke test to catch regressions in subcommand help messages.
- Standardize and clean up logging and error handling for better maintainability.

KushalMeghani1644 committed:
Fix CI failing

KushalMeghani1644 committed:
Merge branch 'dev' into feat

KushalMeghani1644 committed:
Merge branch 'feat' of github.com:KushalMeghani1644/tauri into feat

KushalMeghani1644 committed:
Merge branch 'dev' into feat

KushalMeghani1644 committed:
Merge branch 'dev' into feat

KushalMeghani1644 committed:
Merge branch 'dev' into feat

KushalMeghani1644 committed:
Merge branch 'dev' into feat

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -115,7 +106,7 @@ impl Display for RunMode {
 }
 
-#[derive(Deserialize)]
+#[derive(Deserialize, Default)]
```
Is that used anywhere? Same for PackageJson

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -178,6 +169,13 @@ fn format_error<I: CommandFactory>(err: clap::Error) -> clap::Error {
   err.format(&mut app)
 }
 
+fn get_verbosity(cli_verbose: u8) -> u8 {
```
> Verbosity consistency: Introduced a helper function to centralize verbosity parsing from environment variables and CLI flags, ensuring predictable logging levels across all subcommands.

This function is only used once, where the code was inlined before. I don't mind having this in a separate function but wanted to mention this.

KushalMeghani1644 replied:
I made this function because the logic may be reused (e.g. if we later need verbosity in other commands),
but I’m fine inlining it if preferred.

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -332,18 +315,25 @@ impl CommandExt for Command {
+      .get_args()
+      .map(|a| a.to_string_lossy())
+      .collect::<Vec<_>>()
+      .join(" ");
```
i wonder if that's slower than fold but i agree that it's easier on the eyes.

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -186,10 +184,6 @@ fn format_error<I: CommandFactory>(err: clap::Error) -> clap::Error {
 /// 1. `tauri-cli 1 2 3` -> `1 2 3`
 /// 2. `cargo tauri 1 2 3` -> `1 2 3`
 /// 3. `node tauri.js 1 2 3` -> `1 2 3`
-///
-/// The passed `bin_name` parameter should be how you want the help messages to display the command.
-/// This defaults to `cargo-tauri`, but should be set to how the program was called, such as
-/// `cargo tauri`.
```
Were these comments incorrect?

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -359,16 +349,17 @@ impl CommandExt for Command {
-            lines.extend(line.as_bytes().to_vec());
+            if let Ok(mut lines) = stdout_lines_.lock() {
+              lines.extend(line.as_bytes());
+            }
```
i think we should keep the lock() where it was for performance reasons. the next time we will need a lock is after these threads are stopped.

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -359,16 +349,17 @@ impl CommandExt for Command {
+            }
           }
-          Err(_) => (),
+          Err(_) => break,
```
probably not a good idea, just because one line may error out doesn't mean we should stop reading completely. Except of course if we've had issues with the old behavior which i'm not aware of.

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -378,16 +369,17 @@ impl CommandExt for Command {
+            }
           }
-          Err(_) => (),
+          Err(_) => break,
```
same

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -423,4 +415,10 @@ mod tests {
     Cli::command().debug_assert();
   }
+
+  #[test]
+  fn help_output_includes_build() {
+    let help = Cli::command().render_help().to_string();
+    assert!(help.contains("Build"));
+  }
```
i'm kinda failing to see the value of this test but it's not technically wrong so fair enough i guess

FabianLars commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -65,28 +65,19 @@ impl FromStr for ConfigValue {
-            format!("failed to parse config at {} as JSON5", path.display())
-          })?))
-        }
-        // treat all other extensions as json
-        _ => Ok(Self(
-          // from tauri-utils/src/config/parse.rs:
-          // we also want to support **valid** json5 in the .json extension
-          // if the json5 is not valid the serde_json error for regular json will be returned.
```
since the logic is still the same we should probably keep the comments as well

FabianLars replied:
@KushalMeghani1644 can you copy these comments over as well?

KushalMeghani1644 replied:
Oh yes! I missed these! ~~though~~ Now I have added them back now. Sorry for the inconsistency

KushalMeghani1644 committed:
Merge branch 'dev' into feat

KushalMeghani1644 committed:
Adress review

KushalMeghani1644 committed:
Merge branch 'feat' of github.com:KushalMeghani1644/tauri into feat

KushalMeghani1644 commented:
Hey! @FabianLars I have made the changes, you reviewed/requested... I have added back the comments that I had blindly removed... removed the `Default` as-well. I hope the changes made are fine. :)

KushalMeghani1644 committed:
Add comments

FabianLars reviewed (approved):
thanks :)

Legend-Master commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -222,23 +224,17 @@ where
-    .ok()
-    .and_then(|v| v.parse().ok())
-    .unwrap_or(cli.verbose);
-  // set the verbosity level so subsequent CLI calls (xcode-script, android-studio-script) refer to it
```
Can we add back those comments? Not sure why you removed them

FabianLars replied:
I could swear all your marked comments were added back... Maybe it's time for me to retire lol

KushalMeghani1644 replied:
Oops! I overlooked these comments, I shall stop using neovim... Jokes aside, I will have them added back through another Pr (since this is merged).

Legend-Master commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -222,23 +224,17 @@ where
+  if let Err(err) = builder
     .format_indent(Some(12))
     .filter(None, verbosity_level(verbosity_number).to_level_filter())
-    // golbin spams an insane amount of really technical logs on the debug level so we're reducing one level
```
same here

Legend-Master commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -222,23 +224,17 @@ where
       Some("goblin"),
       verbosity_level(verbosity_number.saturating_sub(1)).to_level_filter(),
     )
-    // handlebars is not that spammy but its debug logs are typically far from being helpful
```
same here

Legend-Master commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -321,8 +314,6 @@ fn prettyprint_level(lvl: Level) -> &'static str {
 }
 
 pub trait CommandExt {
-  // The `pipe` function sets the stdout and stderr to properly
```
same here
